libpod/adapter/runtime_remote.go

```diff
@@ -320,6 +322,129 @@ func (r *LocalRuntime) Config(name string) *libpod.ContainerConfig {
 	return &data
 
 }
+
+// PruneImages is the wrapper call for a remote-client to prune images
 func (r *LocalRuntime) PruneImages(all bool) ([]string, error) {
 	return iopodman.ImagesPrune().Call(r.Conn, all)
 }
+
+// Export is a wrapper to container export to a tarfile
+func (r *LocalRuntime) Export(name string, path_ string) error {
+	tempPath, err := iopodman.ExportContainer().Call(r.Conn, name, "")
+	if err != nil {
+		return err
+	}
+
+	outputFile, err := os.Create(path_)
+	if err != nil {
+		return err
+	}
+	defer outputFile.Close()
+
+	writer := bufio.NewWriter(outputFile)
+	defer writer.Flush()
+
+	//TODO make this something for reals
+	reply, err := iopodman.ReceiveFile().Send(r.Conn, varlink.Upgrade, tempPath, true)
+	if err != nil {
+		return err
+	}
+
+	length, _, err := reply()
+	if err != nil {
+		return err
+	}
+
+	buf4096 := make([]byte, 4096)
+	var buf []byte
+
+	toBeCopied := int(length)
+
+	for toBeCopied > 0 {
+		if toBeCopied >= 4096 {
+			buf = buf4096
+		} else {
+			buf = make([]byte, toBeCopied)
+		}
+
+		numBytesRead, err := r.Conn.Reader.Read(buf)
+
+		if numBytesRead == toBeCopied {
+			err = nil
+		}
+
+		if err != nil {
+			break
+		}
+
+		if numBytesRead == 0 {
+			break
+		}
+
+		toBeCopied = toBeCopied - numBytesRead
+
+		numBytesToWrite := numBytesRead
+
+		for i := 0; numBytesToWrite > 0; {
+			w, err := writer.Write(buf[i:(i + numBytesToWrite)])
+			if err != nil {
+				fmt.Printf("Error: SendFile() Write() %v\n", err)
+			}
+			i = i + w
+			numBytesToWrite = numBytesToWrite - w
+		}
+
+		if numBytesToWrite > 0 || err != nil {
+			break
+		}
+	}
+	if toBeCopied > 0 || err != nil {
+		//os.Remove(file_handle)
+		return err
+	}
+
+	return nil
+
+}
+
+// Import implements the remote calls required to import a container image to the store
+func (r *LocalRuntime) Import(ctx context.Context, source, reference string, changes []string, history string, quiet bool) (string, error) {
+	// First we send the file to the host
+	fs, err := os.Open(source)
+	if err != nil {
+		return "", err
+	}
+
+	fileInfo, err := fs.Stat()
+	if err != nil {
+		return "", err
+	}
+	reply, err := iopodman.SendFile().Send(r.Conn, varlink.Upgrade, "", int64(fileInfo.Size()))
+	if err != nil {
+		return "", err
+	}
+	_, _, err = reply()
+	if err != nil {
+		return "", err
+	}
+
+	reader := bufio.NewReader(fs)
+	_, err = reader.WriteTo(r.Conn.Writer)
+	if err != nil {
+		return "", err
+	}
+	r.Conn.Writer.Flush()
+
+	// All was sent, wait for the ACK from the server
+	tempFile, err := r.Conn.Reader.ReadString(':')
```

Comment:
Is there a timer in ReadString() or are we waiting forever?

Reply:
forever